docs/dcp2_system_design.rst

```diff
@@ -0,0 +1,1465 @@
+       must record an error if it detects more than one object with the
+       same ``links/{links_id}_{version}\_`` prefix.
+
+.. [7]
+   The staging area URI is guaranteed to end in a slash.
+
+.. [8]
+   The entity ID can also be found in each document under
+   ``.provenance.document_id``.
+
+Data file paths
+---------------
+
+Metadata entities that are instances of the `HCA schema types for
+describing data files`_
+refer to their corresponding data files using the
+`file_core.file_name`_
+property. This property contains a file path, not a unique identifier.
+In DCP/1, that path to a data file is interpreted relative to the bundle
+referencing the data file. That's actually the only means in DCP/1 by
+which a data file is connected to the metadata file describing it.
+
+.. _HCA schema types for describing data files: https://github.com/HumanCellAtlas/metadata-schema/tree/master/json_schema/type/file
+.. _file_core.file_name: https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/core/file/file_core.json#L24
+
+In the DCP/2 MVP, import adapters instead create *file descriptors* that
+connect metadata entities to the data files they describe. A file
+descriptor is an object in the staging area, whose name contains the
+metadata entity's coordinates (``entity_type``,
+``entity_id``, and ``version``). The descriptor contains
+the name of the data file relative to the staging area as well as the
+data file's unique identifier and version.
+
+The importer is free to use any virtual file path for TDR it deems
+appropriate. »Possible virtual file path patterns are
+*content-addressed* (using just ``{descriptor.sha1}`` for example) or
+*subgraph-relative* (using
+``{links_id}/{content.file_core.file_name}``).«
+
+Initially, the bundle-relative path of a data file in DCP/1 was not
+allowed to contain slashes, forcing bundles to be flat. This limitation
+has been in part addressed, but when the DCP/1 was shut down, many
+bundles (like those with zarray matrices) still contained file names
+that work around this limitation by using ``!`` instead of ``/``. The DSS
+adapter undoes the substitution. That is the only modification the DSS
+adapter applies to the `file_core.file_name`_ property.
+
+Importer idempotence
+--------------------
+
+The importer should not copy a data file if it is already present in TDR
+and the checksums match between the copy of the file in the staging area
+and the one in TDR/Firestore.
+
+Similarly, the importer must not create a new row in a TDR table if that
+row would be identical to another row except for its ``rowid``.
+
+Summary of schema changes
+-------------------------
+
+»This section lists the schema changes that were made as part of the MVP
+release of the DCP/2. The table is for information only.«
+
+.. list-table::
+
+    * - file_descriptor.json
+      - `metadata-schema #1289`_
+        `metadata-schema #1302`_
+        `metadata-schema #1317`_
```

Comment:
These links display space separated. Is there any way to make them display newline separated like in the original doc?

Reply:
Yes, great suggestion.